notebooks/ch17/classification_precision_recall.md

```diff
@@ -0,0 +1,327 @@
+
+The **classification threshold** is a value that determines what class a data point is assigned to; points that fall on opposite sides of the threshold are labeled with different classes. Recall that logistic regression outputs a probability that the data point belongs to the positive class. If the probability output on $x_1$ is .90 and the output on $x_2$ is .1, then we can be fairly confident that $x_1$ is spam and $x_2$ is ham. However, if the probability associated with $x_3$ is around .55, there is greater uncertainty about the correct class. If the classification threshold is .50, $x_3$ is labeled as spam, but if the threshold is .60, $x_3$ is labeled as ham. Depending on the classification problem, false positives and false negatives may have different importance. Set the classification threshold based on which error type you want to minimize: minimizing false positives maximizes precision and minimizing false negatives maximizes recall. For email classification, false positives result in important emails being filtered out, so they are worse than false negatives, in which a spam email winds up in the inbox. In medical settings, however, false negatives in a diagnostic test can be much more consequential than false positives.
+
+A **precision-recall curve** plots the precision and recall at all classification thresholds. A point $(r, p)$ on the curve represents the recall and precision associated with a specific threshold value. The precision-recall curve for `words_list_model` is shown below:
+
+
+```python
+from sklearn.metrics import precision_recall_curve
+
+
+probabilities = words_list_model.predict_proba(X_train)[:, 1]
+precision, recall, _ = precision_recall_curve(y_train, probabilities)
+plt.step(recall, precision, color='b', alpha=0.2,
+         where='post')
+plt.fill_between(recall, precision, step='post', alpha=0.2,
+                 color='b')
+plt.xlabel('Recall')
+plt.ylabel('Precision')
+plt.title('Spam vs. Ham Classifier Precision-Recall Curve')
+```
+
+
+
+
+    array([ 0.02,  0.03,  0.05,  0.05,  0.06,  0.07,  0.11,  0.13,  0.14,
+            0.17,  0.24,  0.26,  0.3 ,  0.33,  0.48,  0.56])
+
+
+
+
+![png](classification_precision_recall_files/classification_precision_recall_17_1.png)
+
+
+See [here](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_recall_curve.html#sklearn.metrics.precision_recall_curve) for scikit-learn's precision-recall function.
+
+The **Area Under the Curve (AUC)** is a single number performance summary of the classifier. A large AUC means that the model has both high precision and recall, which implies that it has few false postivives and false negatives. The AUC for `words_list_model` is:
```

Comment:
We should add a sentence here explaining how to calculate the AUC, what an AUC of 1 means, and how a random classifier performs.

Reply:
I was able to address all your comments except this one (please read my longer comment below).